### PR TITLE
Reduce allocations in Microsoft.Extensions.Configuration

### PR DESCRIPTION
**Pull request by pentp**
Currently calculating paths always allocates a `string[]` just to concatenate 3 strings.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 1]
```diff
@@ -381,7 +384,7 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
                             (interfaceGenericType == typeof(ICollection<>) || interfaceGenericType == typeof(IList<>)))
                         {
                             // For ICollection<T> and IList<T> we bind them to mutable List<T> type.
-                            Type genericType = typeof(List<>).MakeGenericType(type.GenericTypeArguments[0]);
+                            Type genericType = typeof(List<>).MakeGenericType(type.GenericTypeArguments);
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:387]: > Type genericType = typeof(List<>).MakeGenericType(type.GenericTypeArguments); [](http://example.com/codeflow?start=27&length=78)

why this change? this doesn't look right to me. `type.GenericTypeArguments` will return array of the type and not just the type itself.

Comment (pentp) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:387]: It will return an array of type arguments for that type (with a single element) and `MakeGenericType` takes an array of type arguments (using `params` arrays), so this just avoids an array allocation here.

**Comment by tarekgh** [comment]
@pentp did you run into any allocation problem to provide such fixes? or you just proactively looking for places we can optimize? Is it possible to provide some allocation numbers to see if it is worth optimizing? We are trying to be careful touch the configuration binding code as we experienced some regressions before.

**Comment by pentp** [comment]
> @pentp did you run into any allocation problem to provide such fixes? or you just proactively looking for places we can optimize?

I just noticed that the configuration key (path) generation code was really wasteful when I was adopting a new configuration loading mechanisms for our app, but didn't gather any numbers.

**Review by tarekgh — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationPath.cs [hunk 2]
```diff
@@ -54,7 +54,7 @@ public static string Combine(IEnumerable<string> pathSegments)
                 return path;
             }
 
-            int lastDelimiterIndex = path.LastIndexOf(KeyDelimiter, StringComparison.OrdinalIgnoreCase);
+            int lastDelimiterIndex = path.LastIndexOf(':');
             return lastDelimiterIndex == -1 ? path : path.Substring(lastDelimiterIndex + 1);
```
Comment (gfoidl) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationPath.cs:58]:
```suggestion
            return lastDelimiterIndex < 0 ? path : path.Substring(lastDelimiterIndex + 1);
```
as micro-optimization. Comparing to `0` is a little tad cheaper at cpu-level (comparison and the jump can be fused together).

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationPath.cs [hunk 3]
```diff
@@ -70,7 +70,7 @@ public static string Combine(IEnumerable<string> pathSegments)
                 return null;
             }
 
-            int lastDelimiterIndex = path.LastIndexOf(KeyDelimiter, StringComparison.OrdinalIgnoreCase);
+            int lastDelimiterIndex = path.LastIndexOf(':');
             return lastDelimiterIndex == -1 ? null : path.Substring(0, lastDelimiterIndex);
```
Comment (gfoidl) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationPath.cs:74]:
```suggestion
            return lastDelimiterIndex < 0 ? null : path.Substring(0, lastDelimiterIndex);
```

File: src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationSection.cs [hunk 4]
```diff
@@ -69,11 +69,11 @@ public ConfigurationSection(IConfigurationRoot root, string path)
         {
             get
             {
-                return _root[ConfigurationPath.Combine(Path, key)];
+                return _root[Path + ConfigurationPath.KeyDelimiter + key];
```
Comment (gfoidl) [inline src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationSection.cs:72]: Here are three places where the lookup-key is created. Although it's trivial code, should this be moved to a helper method in order to have only one place where that lookup-key is built?

Comment (pentp) [reply src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationSection.cs:72]: Ideally, should add an overload to `ConfigurationPath.Combine` that takes two path strings (the most common case), but adding it as an internal API wouldn't work because it's in a different assembly and adding a public API needs the full API review process. I'm willing to create an API review request for this, if it sounds useful enough?

Comment (gfoidl) [reply src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationSection.cs:72]: I don't think a public API is needed here. I thought just a helper in `ConfigurationSection` (as the PR's diff shows three places where that could be used).
